File: algo-armstrong-numbers/python/armstrong_numbers.py
#!/usr/bin/env python3.12
import re

# How can you make this more scalable and reusable later?

#Begin Helper Functions

def split_digits(int_num) :
    """This function splits the digits of a number into a list

    Args:
        int_num (int) Number to be split
    Returns:
        (list) List of digits from int_num
    """
    lst_result = []

    # Iterate through the number (as a string) and match each digit
    for str_digit in re.findall(r'\d', f"{int_num}") :
        lst_result.append(str_digit)

    return lst_result

def calculate_number(lst_digits) :
    """This function calculates the total of each digit of a list calculated to the power of the list length

    Args:
        lst_digits (list) List of digits to calculate
    Returns:
        (int) Calculated total of digits to the power of the length of the list
    """
    int_total = 0
    lst_subtotals = []

    # Iterate through lst_digits and calculate the digits to the power of the length of the list
    for int_digit in lst_digits :
        int_this_subtotal = pow(int(int_digit), len(lst_digits))

        # Add this subtotal to the list of subtotals
        lst_subtotals.append(int_this_subtotal)

    # Calculate the sum of all of the subtotals
    # TODO: Can this use Reduce?
    for int_digit in lst_subtotals :
        int_total += int_digit

    return int_total

# End Helper Functions

def find_armstrong_numbers(numbers):
    """This function accepts a range of numbers as a list and returns the Armstrong Numbers found in that list

    Args:
        numbers - (list) A range of numbers to search
    Returns:
        (list) List containing Armstrong Numbers found in numbers
    """
    lst_result = []

    for int_num in numbers :
        # TODO: Why did I make this variable?
        int_this_num = int_num

        # split each of the digits of this number into a list
        lst_this_num_digits = split_digits(int_this_num)

        # Calculate the total of this digit to the power of the list length
        int_this_total = calculate_number(lst_this_num_digits)  

        # Compare the previous total to the current number
        if int_this_total == int_this_num :
            lst_result.append(int_this_num)
    
    return lst_result



#print(find_armstrong_numbers([5,371,333]))

File: algo/algo-armstrong-numbers/js/armstrongNumbers.js
#!/usr/bin/env node
// How can you make this more scalable and reusable later?

// Begin Helper Functions

/**
 * This function splits the digits of a number into an array.
 * @param {number} intNum Number to be split
 * @returns Array of digits from intNum
 */
function splitDigits(intNum) {

    let arrResult = [];

    // Iterate through the number (as a string) and match each digit (globally)
    for(let strDigit of `${intNum}`.matchAll(/\d/g)) {
        arrResult.push(strDigit[0])
    }
    return arrResult;
}

/**
 * This function calculates the total of each digit as an array calculated to the power
 * of the array length.
 * @param {array} arrNum Array of digits to calculate
 * @returns Calculated total of digits to the power of the length of the given array
 */
function calculateNumber(arrNum) {
    let intTotal = 0;
    let arrSubTotals = [];

    // Iterate through arrNum and calculate the digits to the power of the length of the array
    for(let intDigit of arrNum) {
        let intThisSubTotal = (intDigit ** arrNum.length);
        // Add them to the subtotal array 
        arrSubTotals.push(intThisSubTotal);
    }

    // Add all of the subtotals together
    for(let intDigit of arrSubTotals) {
        intTotal += intDigit;
    }

    return intTotal;

}

// Copied from armstrongNumbersSpec.js to use for internal testing
function createArrayOfNum(maxValue) {
    return [...Array(maxValue).keys()];
  }

// End Helper Functions

/**
 * This function accepts a range of numbers as an array and returns the Armstrong Numbers
 * found in that array
 * @param {array} arrInput Array of a range of numbers to search for Armstrong Numbers
 * @returns Array containing Armstrong Numbers found in the given range
 */
function findArmstrongNumbers(arrInput) {
    let arrResult = [];

    for(let intNumber of arrInput) {
        // Take each number from the array and turn it into a string
        // Iterate through the characters of the string to calculate
        // the number to determine if it is an Armstrong Number or not

        let intThisNum = intNumber;

        let arrThisNumDigits = splitDigits(intThisNum)

        let intThisTotal = calculateNumber(arrThisNumDigits)

        if(intThisTotal === intNumber) {
            arrResult.push(intNumber)
        }
    }

    return arrResult;

}

exports.findArmstrongNumbers = function() {
//exports.findArmstrongNumbers;
};

//module.exports = findArmstrongNumbers;

//console.log(findArmstrongNumbers(createArrayOfNum(99)))